lib/puppet/type.rb

```diff
@@ -1205,15 +1205,16 @@ def self.instances
       provider.instances.collect do |instance|
         # We always want to use the "first" provider instance we find, unless the resource
         # is already managed and has a different provider set
-        other = provider_instances[instance.name]
+        title = instance.respond_to?(:title) ? instance.title : instance.name
```

Comment:
Just adding some context about name vs title. So `resource.name` came first and it is always `[:name]`:
https://github.com/puppetlabs/puppet/blob/f3265237264de95d59807ad9333182b5e405f9b9/lib/puppet/type.rb#L2583-L2590

The `Type#[]` method looks up the `:name` as a parameter or property https://github.com/puppetlabs/puppet/blob/f3265237264de95d59807ad9333182b5e405f9b9/lib/puppet/type.rb#L645-L650

At a later point, namevars were added. There are a few different ways to declare a namevar. See https://github.com/puppetlabs/puppet-specifications/blob/master/language/resource_types.md#isnamevar.

1. Use the `isnamevar` DSL method
1. Pass `namevar: true` as an option when declaring the parameter, eg `newparam(:name, :namevar => true)`
1. Parameters named `:name` are implicit namevars.

Composite namevars were added in https://github.com/puppetlabs/puppet/commit/7c6b8836453b2b1e8679923f98854be3b0022edd.

Prior to that the title was looked up from the first namevar parameter or property:
https://github.com/puppetlabs/puppet/commit/607d7c6afec7ab73203382676cc98b6d52898e1d#diff-7f00f7dbc0f9bfecbaadfec966d0b8b1R1858-R1883. From what I can tell the fallback case where `@title` is nil is dead code, but I'm not sure...

It's be good to verify this PR using types that have:

#. 1 explicit namevar
#. an implicit `:name` namevar
#. N explicit namevars
#. a namevar that isn't `:name`, like the `file` type uses `:path`.

Reply:
@joshcooper Thanks for digging this stuff up.

I believe this part of the code only runs when collecting resources (like `puppet resource something` or when setting `purge => true`), so it should not impact use cases where we only ensure specific resources.

1. From what I see, most `:name` params also use `isnamevar`, this is the most common case (`service`, `user`)
2. Here we can just remove `isnamevar` from one of the providers above, things still work the same
3. `package` and `sshkey` have 2 namevars. The `package` provider still doesn't show packages installed by multiple providers since they all have the same title. In theory, if we change the resources' title to something like `package@provider` this should work here as well... but it's not in the scope of this PR.
4. `file` instances cannot be collected, you can only run `puppet resource` on a specific file/directory, and it works the same. I found the `registry` module that has the `registry_key` type with `:path` as namevar https://github.com/puppetlabs/puppetlabs-registry/blob/master/lib/puppet/type/registry_key.rb#L39. You are able to run `puppet resource registry_key` and it shows only the top-level keys, not sure if the module was meant to be used this way.

I validated this as much as I could by hand. I ran the `puppetlabs-registry` acceptance tests with and without this PR. I also started an ad-hoc puppet-agent pipeline that runs all tests, not just the high risk ones, to ensure nothing broke. We may see unrelated tests failing due to them not being run for some time.